### PR TITLE
Add occurrences for classOf

### PR DESCRIPTION
**Pull request by sswistun-vl**
Before, classOf didn't generate occurrences, both for `classOf` method.
Now, it generates them as `scala.Predef.classOf()`. Tests were adjusted to reflect the changes.

Related to https://github.com/scalameta/scalameta/issues/1909

### CODE REVIEW

**Comment by tgodzik** [comment]
@olafurpg it seems the classOf[A] is a Literal inside the compiler and there is no way to get the possitions of symbols aside from doing it manually. Let us know if there is any better solution for this!

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala [hunk 1]
```diff
@@ -636,7 +636,8 @@ trait TextDocumentOps { self: SemanticdbOps =>
                 traverse(gtree.check())
 
               case gtree: g.Literal
-                  if gtree.tpe != null && gtree.tpe.typeSymbol.name.decode.contentEquals("Class") =>
+                  if gtree.tpe != null && gtree.tpe.typeSymbol.name.decode
```
Comment (olafurpg) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: The idiomatic way to compare symbols is via `==`, you can get a reference to the `java.lang.Class` symbol in `g.definitions.ClassClass`

Screenshot 2019-11-15 at 09 03 40

Comment (olafurpg) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Can we add an additional guard that the source code actually contains `classOf`?

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala [hunk 2]
```diff
@@ -635,6 +635,21 @@ trait TextDocumentOps { self: SemanticdbOps =>
...
+                val mpos = gtree.pos.toMeta
+                val mposFix = new m.Position.Range(mpos.input, mpos.start, mpos.start + coLen)
+
+                occurrences(mposFix) = "scala/Predef.classOf()."
+                binders += mposFix
+
+                val cpos =
+                  new m.Position.Range(mpos.input, mpos.start + coLen + 1, mpos.end - 1)
+                occurrences(cpos) = gtree.value.typeValue.typeSymbol.toSemantic
+                binders += cpos
```
Comment (olafurpg) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Is this line needed? `classOf` is not a definition

Comment (olafurpg) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: same for `mposFix`

File: tests/jvm/src/test/resources/example/InstrumentTyper.scala [hunk 3]
```diff
@@ -23,5 +23,5 @@ class InstrumentTyper/*<=example.InstrumentTyper#*/ { self/*<=local0*/: AnyRef =
   def existential/*<=example.InstrumentTyper#existential().*/: U/*=>local1*/[Int/*=>scala.Int#*/] forSome { type U/*<=local1*/[T/*<=local2*/ <: Int] } = ???/*=>scala.Predef.`???`().*/
   type AnnotatedType/*<=example.InstrumentTyper#AnnotatedType#*/ = Int/*=>scala.Int#*/ @param
   def singletonType/*<=example.InstrumentTyper#singletonType().*/(x/*<=example.InstrumentTyper#singletonType().(x)*/: Predef.type) = ???/*=>scala.Predef.`???`().*/
-  final val clazzOf/*<=example.InstrumentTyper#clazzOf.*/ = classOf[Option[Int]]
+  final val clazzOf/*<=example.InstrumentTyper#clazzOf.*/ = classOf/*<=scala.Predef.classOf().*/[Option[Int]/*<=scala.Option#*/]
```
Comment (olafurpg) [inline tests/jvm/src/test/resources/example/InstrumentTyper.scala]: Is there a way to also capture the reference to `Int` inside the type argument?

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala (hunk 1 above)
Comment (sswistun-vl) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Fixed.

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala (hunk 2 above)
Comment (sswistun-vl) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Fixed.

**Comment by tgodzik** [comment]
We can either merge this (will work only with the first class not ones with type params, so not ideal) and work on fixing the compiler separately OR drop it entirely for know. I am on the edge for making it partially here, but it would help with most use cases in Metals.

@olafurpg what do you think?

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala [hunk 4]
```diff
@@ -604,6 +604,13 @@ trait TextDocumentOps { self: SemanticdbOps =>
               case _ =>
             }
 
+            def isClassOf(pos: m.Position): Boolean = {
+              val (chars, classOfChars) = (pos.input.chars, "classOf".toCharArray)
```
Comment (olafurpg) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: nit: a more idiomatic way to declare two variables
```suggestion
              val chars = pos.input.chars
              val classOfChars = "classOf".toCharArray
```

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala [hunk 5]
```diff
@@ -604,6 +604,13 @@ trait TextDocumentOps { self: SemanticdbOps =>
...
+
+              chars.length >= (classOfChars.length + pos.start) && (0 until classOfChars.length)
```
Comment (olafurpg) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Nit: please insert manual line breaks around infix operators to make it clearer to read the conditions 
```suggestion
              chars.length >= (classOfChars.length + pos.start) &&
                (0 until classOfChars.length)
```

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala [hunk 6]
```diff
@@ -635,6 +642,18 @@ trait TextDocumentOps { self: SemanticdbOps =>
               case gtree: g.TypeTreeWithDeferredRefCheck =>
                 traverse(gtree.check())
 
+              case gtree: g.Literal
+                  if isClassOf(gtree.pos.toMeta) && gtree.tpe != null && gtree.tpe.typeSymbol == g.definitions.ClassClass && gtree.pos.isRange =>
```
Comment (olafurpg) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: nit: multiple lines make it easier to read the multiple conditions
```suggestion
                  if isClassOf(gtree.pos.toMeta) &&
                    gtree.tpe != null &&
                    gtree.tpe.typeSymbol == g.definitions.ClassClass &&
                    gtree.pos.isRange =>
```

File: tests/jvm/src/test/resources/example/Types.scala [hunk 7]
```diff
@@ -111,6 +111,6 @@ object Test/*<=types.Test.*/ {
     final val bool/*<=types.Test.Literal.bool.*/ = true
     final val unit/*<=types.Test.Literal.unit.*/ = ()
     final val javaEnum/*<=types.Test.Literal.javaEnum.*/ = java.nio.file.LinkOption/*=>java.nio.file.LinkOption#*/.NOFOLLOW_LINKS/*=>java.nio.file.LinkOption#NOFOLLOW_LINKS.*/
-    final val clazzOf/*<=types.Test.Literal.clazzOf.*/ = classOf[Option[Int]]
+    final val clazzOf/*<=types.Test.Literal.clazzOf.*/ = classOf/*=>scala.Predef.classOf().*/[Option[Int]/*=>scala.Option#*/]
```
Comment (olafurpg) [inline tests/jvm/src/test/resources/example/Types.scala]: I would expect the occurrence position to enclose the `Option` identifier 🤔 
```suggestion
    final val clazzOf/*<=types.Test.Literal.clazzOf.*/ = classOf/*=>scala.Predef.classOf().*/[Option/*=>scala.Option#*/[Int]]
```

File: tests/jvm/src/test/resources/metac.expect [hunk 8]
```diff
@@ -1801,6 +1801,8 @@ Occurrences:
 [24:20..24:21): x <= example/InstrumentTyper#singletonType().(x)
 [24:38..24:41): ??? => scala/Predef.`???`().
 [25:12..25:19): clazzOf <= example/InstrumentTyper#clazzOf.
+[25:22..25:29): classOf => scala/Predef.classOf().
+[25:30..25:41): Option[Int] => scala/Option#
```
Comment (olafurpg) [inline tests/jvm/src/test/resources/metac.expect]: This means that renaming `Option` to `Foobar` would replace `Option[Int]` with `Foobar`

File: tests/jvm/src/test/resources/example/Types.scala (hunk 7 above)
Comment (olafurpg) [reply tests/jvm/src/test/resources/example/Types.scala]: Another test case to handle: qualifiers
```scala
classOf[scala.Option[Int]]
```

Comment (olafurpg) [reply tests/jvm/src/test/resources/example/Types.scala]: Another test case: renamed imports
```scala
import scala.concurrent.{Future => Fut}

classOf[Fut[Int]]
```

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala (hunk 5 above)
Comment (sswistun-vl) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Fixed.

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala (hunk 6 above)
Comment (sswistun-vl) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Fixed.

File: tests/jvm/src/test/resources/metac.expect (hunk 8 above)
Comment (sswistun-vl) [reply tests/jvm/src/test/resources/metac.expect]: Fixed. You might want to check out this solution as it's based on character occurrence in string - I'm not sure if it's a proper way to do it.

File: tests/jvm/src/test/resources/example/Types.scala (hunk 7 above)
Comment (sswistun-vl) [reply tests/jvm/src/test/resources/example/Types.scala]: I've added tests, but not sure if they are in the right place.

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala (hunk 4 above)
Comment (sswistun-vl) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala]: Fixed.

**Comment by tgodzik** [comment]
I think this is a bit of a dead end - we should revert to just including the classOf occurence and aim to fix the exact positions in the compiler. @sswistun-vl

**Comment by sswistun-vl** [comment]
I've removed occurrences for classOf arguments as they can't be resolved properly right now.

**Comment by tgodzik** [comment]
@olafurpg we reverted the part reposible for the type param and I think this is mergable now. It will at least fix the classOf occurences and provide an entry point when we fix the possitions for the type params inside classOf. What do you think?

**Comment by tgodzik** [comment]
> LGTM

Thanks for taking a look at it!

**Pull request merged**